### PR TITLE
[core][spark] Supports to push down limit

### PR DESCRIPTION
**Pull request by YannByron**
### Purpose
Support to push down limit to accelerate query.

Linked issue: close #xxx

### CODE REVIEW

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  override def pruneColumns(requiredSchema: StructType): Unit = {
+    val pruneFields = requiredSchema.fieldNames
+    val fieldNames = table.rowType.getFieldNames
+    val projected = pruneFields.map(field => fieldNames.indexOf(field))
+    this.projectedIndexes = Some(projected)
+  }
+
+  override def pushLimit(limit: Int): Boolean = {
+    if (table.isInstanceOf[AppendOnlyFileStoreTable]) {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala:81]: No need to `if append table`? Paimon-core scan should take care about this.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  override def pruneColumns(requiredSchema: StructType): Unit = {
+    val pruneFields = requiredSchema.fieldNames
+    val fieldNames = table.rowType.getFieldNames
+    val projected = pruneFields.map(field => fieldNames.indexOf(field))
+    this.projectedIndexes = Some(projected)
+  }
+
+  override def pushLimit(limit: Int): Boolean = {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala:80]: Does Spark SQL will not push limit if there is filter?

File: paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/PaimonSQLPerformanceTest.scala [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import org.apache.paimon.table.source.DataSplit
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.connector.read.{ScanBuilder, SupportsPushDownLimit}
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+import org.junit.jupiter.api.Assertions
+
+import scala.collection.JavaConverters._
+
+class PaimonSQLPerformanceTest extends PaimonSparkTestBase {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/PaimonSQLPerformanceTest.scala:30]: What this class for? It looks like we can not find performance result from these tests?

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala (hunk 1 above)
Comment (YannByron) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala:81]: Here, a `boolean` value is needed, which is used in SparkSQL. Or, paimon-core needs to provide an api to be called.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala (hunk 2 above)
Comment (YannByron) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala:80]: yes. SparkSQL can guarantee this.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala (hunk 1 above)
Comment (JingsongLi) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala:81]: We can just return false here? Best effort pushdown?

Comment (YannByron) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala:81]: As a temporary solution, it's ok.

**Comment by YannByron** [comment]
These failed UTs are not related to this pr.

**Comment by YannByron** [comment]
https://github.com/apache/incubator-paimon/issues/2404

File: paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java [hunk 4]
```diff
@@ -199,6 +202,11 @@ public FileStoreScan withMetrics(ScanMetrics metrics) {
         return this;
     }
 
+    @Override
+    public FileStoreScan withLimit(int limit) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java:206]: Can limit implementation just in `AppendOnlyFileStoreScan`?

**Comment by JingsongLi** [comment]
Another solution: we can make this limit pushdown generic. We can do this limit in `InnerTableScanImpl`, and we can just check `Split.convertToRawFiles`, check the rowcount, because rawFiles don't need to be merged.

File: paimon-core/src/main/java/org/apache/paimon/table/source/RawFile.java [hunk 5]
```diff
@@ -76,12 +80,18 @@ public long schemaId() {
...
+        return rowCount;
+    }
+
     public void serialize(DataOutputView out) throws IOException {
         out.writeUTF(path);
         out.writeLong(offset);
         out.writeLong(length);
         out.writeUTF(format);
         out.writeLong(schemaId);
+        out.writeLong(rowCount);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/source/RawFile.java:94]: Increment `DataSplit.serialVersionUID` too.

File: paimon-spark/paimon-spark-3.1/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+import org.apache.paimon.table.Table
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.connector.read.{Scan, ScanBuilder, SupportsPushDownFilters, SupportsPushDownRequiredColumns}
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types.StructType
+
+import scala.collection.mutable
+
+class SparkScanBuilder(table: Table)
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-3.1/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala:30]: Can we have a `SparkBaseScanBuilder` to reuse code?

Comment (JingsongLi) [inline paimon-spark/paimon-spark-3.1/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala:30]: You need to add class for spark-3.2 too? And add some itcase for spark-3.2 too (For ITCase, we also can have some abstraction to reduce code replication).

File: paimon-core/src/main/java/org/apache/paimon/table/source/InnerTableScanImpl.java [hunk 7]
```diff
@@ -68,9 +81,96 @@ public TableScan.Plan plan() {
...
+        }
+    }
+
+    private DataSplit composeDataSplit(DataSplit split, long rowCountRequired) {
+        List<DataFileMeta> dataFiles = new ArrayList<>();
+        List<RawFile> rawFiles = new ArrayList<>();
+        long scannedRowCount = 0;
+
+        List<DataFileMeta> originalDataFiles = split.dataFiles();
+        List<RawFile> originalRawFiles = split.convertToRawFiles().get();
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/source/InnerTableScanImpl.java:160]: Just get? Exception when it is empty?

File: paimon-core/src/main/java/org/apache/paimon/table/source/InnerTableScanImpl.java [hunk 8]
```diff
@@ -68,9 +81,96 @@ public TableScan.Plan plan() {
...
+                    break;
+                }
+
+                DataSplit split = splits.get(i);
+                long splitRowCount = getRowCountForSplit(split);
+                if (scannedRowCount + splitRowCount <= pushDownLimit) {
+                    limitedSplits.add(split);
+                    scannedRowCount += splitRowCount;
+                } else {
+                    DataSplit newSplit = composeDataSplit(split, pushDownLimit - scannedRowCount);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/source/InnerTableScanImpl.java:108]: I think we don't need to introduce `composeDataSplit`.

The reason we introduce limit pushdown is to reduce split number. We don't need to reduce files in split.

File: paimon-spark/paimon-spark-3.1/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala (hunk 6 above)
Comment (YannByron) [reply paimon-spark/paimon-spark-3.1/src/main/scala/org/apache/paimon/spark/SparkScanBuilder.scala]: For UT, just copy codes for now. Let refine it in another pr.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**